src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs

```diff
@@ -400,19 +400,39 @@ public void OnCompleted(Func<object, Task> callback, object state)
 
         public void Flush()
         {
-            ProduceStartAndFireOnStarting().GetAwaiter().GetResult();
+            var produceStartTask = ProduceStartAndFireOnStarting();
+            if (produceStartTask.Status != TaskStatus.RanToCompletion)
+            {
+                produceStartTask.GetAwaiter().GetResult();
+            }
+
             SocketOutput.Write(_emptyData);
         }
 
-        public async Task FlushAsync(CancellationToken cancellationToken)
+        public Task FlushAsync(CancellationToken cancellationToken)
+        {
+            var produceStartTask = ProduceStartAndFireOnStarting();
+            if (produceStartTask.Status != TaskStatus.RanToCompletion)
+            {
+                return FlushAsyncAwaited(produceStartTask, cancellationToken);
+            }
+
+            return SocketOutput.WriteAsync(_emptyData, cancellationToken: cancellationToken);
+        }
+
+        public async Task FlushAsyncAwaited(Task produceStartTask, CancellationToken cancellationToken)
```

Comment:
Its delaying invoking the `async` part and calling into a method that the compiler has created as a state machine unless the task has not completed. Similar to directly returning the task where there is only one await; rather than `async`ing the method and awaiting it.

Some places there is no value in this optimization, for example `RequestProcessingAsync` as there is already a statemachine and will always be one and within that context `await` already does this optimization. However in these paths the statemachine creation can be avoided all together.